Feature delete user - check active reservations

Vasilijez opened the pull request:
Closes #6.

First, I'd like to note that the branch name is not ideal, as there's still more work to be done to complete the user deletion feature.

This PR aims to incorporate as many components as feasible within the project's current state.

For more information, check the related issue #6, there will be a series of linked PRs.

__Caveat:__ If you see empty commits, it’s very likely caused by using the `git cherry-pick` command from an old branch and resolving many conflicts.

Part of:
- https://github.com/book-em/reservation-service/pull/16
- https://github.com/book-em/room-service/pull/27
- https://github.com/book-em/user-service/pull/34

Vasilijez committed:
feat: Add `reservationclient` to directly check active guest reservations

Vasilijez committed:
feat: Logged-in user deletes themselves

I also replaced 'Pending' with 'Active' in `GetPendingGuestReservations`, as it seemed to provide more consistency with `GetActiveHostReservations`.

*Tests also need to be updated.

Vasilijez committed:
feat: Update `baseURL` of `RoomClient`

Vasilijez committed:
feat: Finish `GetActiveHostReservations` within `RoomClient`

Vasilijez committed:
feat: Finish `GetActiveHostReservations` within `RoomClient`

Vasilijez committed:
test: Retune unit tests mostly due to `ReservationClient`

Vasilijez committed:
test: Update unit tests for `TestDelete`

Vasilijez committed:
test: Add and update integration tests for `TestIntegration_Delete`

Vasilijez committed:
feat: Update `reservation-client` to use new logging

Vasilijez committed:
feat: Add forgotten error messages in `reservation-client`

Vasilijez committed:
test: Update unit tests with `context.Background()` parameter

Vasilijez committed:
fix: Untar private key for CI

Vasilijez committed:
feat: Use branches with latest changes

I need to use it so the integration tests pass.

magley commented on src/client/reservationclient/model.go:
```diff
@@ -0,0 +1,16 @@
+
+import "time"
+
+type ReservationDTO struct {
+	ID                 uint      `gorm:"primaryKey"`
+	RoomID             uint      `gorm:"not null"`
+	RoomAvailabilityID uint      `gorm:"not null"`
+	RoomPriceID        uint      `gorm:"not null"`
+	GuestID            uint      `gorm:"not null"`
+	DateFrom           time.Time `gorm:"not null"`
+	DateTo             time.Time `gorm:"not null"`
+	GuestCount         uint      `gorm:"not null"`
+	Cancelled          bool      `gorm:"not null"`
+	Cost               uint      `gorm:"not null"`
+}
```
Should a DTO have `gorm` annotations?

Vasilijez replied:
Not really. I accidentally left it there after copying `Reservation` from `model.go` in the `reservation-service`. It’s now removed as part of another issue you pointed out  (https://github.com/book-em/user-service/pull/34/commits/f171d2e01f6d8af17203eabb1e0d7819a1383a6d).

magley commented on src/client/roomclient/model.go:
```diff
@@ -1,7 +1,39 @@
+	GuestCount         uint      `gorm:"not null"`
+	Cancelled          bool      `gorm:"not null"`
+	Cost               uint      `gorm:"not null"`
+}
```
Why does `roomclient` need to define `ReservationDTO`? Just use `reservationclient/model.go`'s definition.

magley replied:
Ok, so I think it's because `getActiveHostReservations` is defined in `roomclient`.

So my question then becomes: why is `room-service` in charge of fetching hosts' reservations? Especially when [the definition of `room-service`'s `GetActiveHostReservations()`](https://github.com/book-em/room-service/pull/27/files#diff-56887d359f3b2e488fce05b5594ee73bb321261ddc658e7901a956fb05edd9d5R745-R777) forwards the request to `reservation-service`.

 Why not do the same here? 

`user-service -> reservation-service`. Skip `room-service` as a middleman. If you need to fetch the rooms before you can get the reservations of a host, do that in `reservation-service`. So the flow would be like:

<img width="1439" height="785" alt="Image" src="https://github.com/user-attachments/assets/5d1ed76b-4f38-4efb-85ed-83abbd9d0b5d" />

Vasilijez replied:
>Ok, so I think it's because getActiveHostReservations is defined in roomclient.

Correct.

> So my question then becomes: why is room-service in charge of fetching hosts' reservations? Especially when [the definition of room-service's GetActiveHostReservations()](https://github.com/book-em/room-service/pull/27/files#diff-56887d359f3b2e488fce05b5594ee73bb321261ddc658e7901a956fb05edd9d5R745-R777) forwards the request to reservation-service.
> 
> Why not do the same here?

That's a good question. It happened because I made a shift during development. Initially, for the host, my idea was to avoid sending too many requests and creating tight coupling, so I wanted to handle everything in a single request. My plan was to contact the room service, check each room for active reservations, and if none had any, delete all rooms. Later, I came across a different design sketch (yours) and shifted from my initial idea, which led to the issue you noticed.

<img width="1574" height="588" alt="image" src="https://github.com/user-attachments/assets/ca67d01d-f37b-43d1-bf8d-67b6ef542125" />

_Sorry for the poor "sequence diagram". It's just a rough sketch I made early on, before any implementation._

I'm not entirely sure what the best approach is now. I'll take some time to think about it, especially since I still need to implement room deletion.

magley replied:
You shouldn't worry too much about sending so many requests because:
1) It's a microservice app, so it "comes with the territory"
2) Since this is just a bunch of GET requests, we won't have any syncing issues (or, at least, none that we have to think about)

My issue with the current implementation is:
1) it's inconsistent between Guest and Host (guest talks to reservation-service directly while host doesn't).
2) reservations are not `room-service`'s concern 
3) there's duplicated logic and code (this might be the worst offender since if we ever change `ReservationDTO`, we have to remember to change it both in `roomclient` and `reservationclient`. If we had a 1000 clients, you'd have to check which ones define `ReservationDTO` every time, as opposed to knowing that `ReservationDTO` is found only in `reservationclient`).
4) implementation details are leaking into other services

I think the `client` part of each microservice should be treated as "private" to _that_ microservice only.

Another way to think about this is: imagine if `room-service` and `reservation-service` were both some external API. Who would you "ask" whether some user has any reservations? As far as `user-service` is concerned, the actual details of how `reservation-service` "answers" that question is not important (black box).

Vasilijez replied:
Okay, I think I got what you meant, I hope I did it correctly this time (https://github.com/book-em/user-service/pull/34/commits/f171d2e01f6d8af17203eabb1e0d7819a1383a6d).

magley reviewed (changes requested):
Ok, two things:

**1) My comments left in this PR review.**

- DTOs should not have gorm annotations since they're not DB models.
- `user-service` should not ask `room-service` to fetch host reservations (it should ask `reservation-service`)

**2) The missing commits.**

I'm guessing this is because `user-service`'s `develop` branch changed in the meantime and you wanted to fetch the new changes before creating this PR?

I had that situation yesterday after rebasing your PR into `develop` and then wanting to create my own PR. I didn't use cherry-picking, but instead:

```
[develop is updated]
git checkout my_branch
git rebase develop
git push --force-with-lease
```

I'm not sure if this would solve your problem, though. I just wanted to get the new commits from develop and then force push my local branch.

Vasilijez committed:
feat: Move logic of `GetActiveHostReservations` to `reservation-service`

Vasilijez commented:
> **2) The missing commits.**
> 
> I'm guessing this is because `user-service`'s `develop` branch changed in the meantime and you wanted to fetch the new changes before creating this PR?
> 
> I had that situation yesterday after rebasing your PR into `develop` and then wanting to create my own PR. I didn't use cherry-picking, but instead:
> 
> ```
> [develop is updated]
> git checkout my_branch
> git rebase develop
> git push --force-with-lease
> ```
> 
> I'm not sure if this would solve your problem, though. I just wanted to get the new commits from develop and then force push my local branch.

I usually follow your approach, but yesterday I intentionally used `git cherry-pick`. All of these changes were on an old branch named `feature-delete-user-finish` (which has been deleted in the meantime). Then I created a new branch, `feature-delete-user-core`, and performed the cherry-pick. I did it because many changes occurred across the entire codebase due to the new logging, and I wanted to avoid rebasing or modifying your commits in history, as that would have caused conflicts in almost every commit.
So for me, it was a better option to go commit by commit and resolve conflicts gradually. It was actually quite simple, but as you can see, a few commits ended up being empty. I assume that's because when resolving conflicts, I discarded all changes from the cherry-picked commit, resulting in no effective changes. I'll just need to be more careful next time not to push empty commits.

magley commented on src/client/roomclient/client.go:
```diff
@@ -1,59 +0,0 @@
-package roomclient
-
-import (
-	utils "bookem-user-service/util"
-	"context"
-	"encoding/json"
-	"fmt"
-	"io"
-	"net/http"
-)
-
-type RoomClient interface {
-	// GetActiveHostReservations finds all reservations made to rooms owned by
-	// `host` that haven't completed yet. The user must be a host.
-	GetActiveHostReservations(ctx context.Context, jwt string) ([]ReservationDTO, error)
-}
-
-type roomClient struct {
-	baseURL string
```
You shouldn't have deleted the entire file `roomclient/client.go`. We'll need `RoomClient` later on, it's ok if it's empty for now.

Vasilijez replied:
https://github.com/book-em/user-service/pull/34/commits/3b867c1afff4dfed0d92d8f66bccb6df23f51604

magley commented:
>I usually follow your approach, but yesterday I intentionally used git cherry-pick. All of these changes were on an old branch named feature-delete-user-finish (which has been deleted in the meantime). Then I created a new branch, feature-delete-user-core, and performed the cherry-pick. I did it because many changes occurred across the entire codebase due to the new logging, and I wanted to avoid rebasing or modifying your commits in history, as that would have caused conflicts in almost every commit.
>So for me, it was a better option to go commit by commit and resolve conflicts gradually. It was actually quite simple, but as you can see, a few commits ended up being empty. I assume that's because when resolving conflicts, I discarded all changes from the cherry-picked commit, resulting in no effective changes. I'll just need to be more careful next time not to push empty commits.

I appreciate the effort but it's caused us more headaches than if you just copy-pasted the code that you need from the old branch. I mean I get that that's not a good idea and we should use cherry-picking but then stuff like this happens.

Vasilijez committed:
feat: Add empty `client.go` file

Vasilijez committed:
feat: Use develop as latest branch from other microservices

This is possible since `room-service` has been discarded (no effective changes), and `reservation-service` has already been merged.

Vasilijez committed:
feat: Inject `RoomClient` in service and tests

magley reviewed (approved):
Good work.
